src/hcpp-hpt.c

```diff
@@ -275,12 +304,163 @@ inline void init_hc_deque_t(hc_deque_t * hcdeq, place_t * pl){
 #endif
 }
 
-/*Initialize deq's buffer (data + capacity)*/
-void init_deq_buffer(hc_workerState * ws, deque_t * deq, int capacity) {
-    //TODO: in our current design this does not makes sense
+#ifdef HC_CUDA
+void *hclib_allocate_at(place_t *pl, size_t nbytes, int flags) {
+    assert(pl); assert(nbytes > 0);
+#ifdef VERBOSE
+    fprintf(stderr, "hclib_allocate_at: pl=%p nbytes=%lu flags=%d, is_cpu? %s",
+            pl, nbytes, flags, is_cpu_place(pl) ? "true" : "false");
+#ifdef HC_CUDA
+    fprintf(stderr, ", is_nvgpu? %s, cuda_id=%d",
+            is_nvgpu_place(pl) ? "true" : "false", pl->cuda_id);
+#endif
+    fprintf(stderr, "\n");
+#endif
+
+    if (is_cpu_place(pl)) {
+#ifdef HC_CUDA
+        if (flags & PHYSICAL) {
+            void *ptr;
+            const cudaError_t alloc_err = cudaMallocHost((void **)&ptr, nbytes);
+            if (alloc_err != cudaSuccess) {
+#ifdef VERBOSE
+                fprintf(stderr, "Physical allocation at CPU place failed with "
+                        "reason \"%s\"\n", cudaGetErrorString(alloc_err));
+#endif
+                return NULL;
+            } else {
+                hclib_memory_tree_insert(ptr, nbytes,
+                        &hcpp_context->pinned_host_allocs);
+                return ptr;
+            }
+        }
+#else
+        HASSERT(flags == NONE);
+#endif
+        return malloc(nbytes);
+#ifdef HC_CUDA
+    } else if (is_nvgpu_place(pl)) {
+        HASSERT(flags == NONE);
+        void *ptr;
+        HASSERT(pl->cuda_id >= 0);
+        CHECK_CUDA(cudaSetDevice(pl->cuda_id));
+        const cudaError_t alloc_err = cudaMalloc((void **)&ptr, nbytes);
+        if (alloc_err != cudaSuccess) {
+#ifdef VERBOSE
+            fprintf(stderr, "Allocation at NVGPU place failed with reason "
+                    "\"%s\"\n", cudaGetErrorString(alloc_err));
+#endif
+            return NULL;
+        } else {
+            return ptr;
+        }
+#endif
+    } else {
+        unsupported_place_type_err(pl);
+        return NULL; // will never reach here
+    }
 }
 
+int is_pinned_cpu_mem(void *ptr) {
+    return hclib_memory_tree_contains(ptr, &hcpp_context->pinned_host_allocs);
+}
+
+void hclib_free_at(place_t *pl, void *ptr) {
+    if (is_cpu_place(pl)) {
+        if (is_pinned_cpu_mem(ptr)) {
+            hclib_memory_tree_remove(ptr, &hcpp_context->pinned_host_allocs);
+            CHECK_CUDA(cudaFreeHost(ptr));
+        } else {
+            free(ptr);
```

Comment:
I don't like this idea of wrapping things in macros, and capitalizing everything. Can you explain your rationale for these changes?